src/avif.c

```diff
@@ -264,45 +273,70 @@ void avifImageSetMetadataXMP(avifImage * image, const uint8_t * xmp, size_t xmpS
     avifRWDataSet(&image->xmp, xmp, xmpSize);
 }
 
-void avifImageAllocatePlanes(avifImage * image, avifPlanesFlags planes)
+avifResult avifImageAllocatePlanes(avifImage * image, avifPlanesFlags planes)
 {
-    int channelSize = avifImageUsesU16(image) ? 2 : 1;
-    int fullRowBytes = channelSize * image->width;
-    int fullSize = fullRowBytes * image->height;
+    if (image->width == 0 || image->height == 0) {
+        return AVIF_RESULT_INVALID_ARGUMENT;
+    }
+    const size_t channelSize = avifImageUsesU16(image) ? 2 : 1;
+    if (image->width > SIZE_MAX / channelSize) {
+        return AVIF_RESULT_INVALID_ARGUMENT;
+    }
+    const size_t fullRowBytes = channelSize * image->width;
+    if ((fullRowBytes > UINT32_MAX) || (image->height > SIZE_MAX / fullRowBytes)) {
+        return AVIF_RESULT_INVALID_ARGUMENT;
+    }
+    const size_t fullSize = fullRowBytes * image->height;
+
     if ((planes & AVIF_PLANES_YUV) && (image->yuvFormat != AVIF_PIXEL_FORMAT_NONE)) {
         avifPixelFormatInfo info;
         avifGetPixelFormatInfo(image->yuvFormat, &info);
 
-        int shiftedW = (image->width + info.chromaShiftX) >> info.chromaShiftX;
-        int shiftedH = (image->height + info.chromaShiftY) >> info.chromaShiftY;
+        // Intermediary computation as 64 bits in case width or height is exactly UINT32_MAX.
+        const uint32_t shiftedW = (uint32_t)((image->width + (uint64_t)info.chromaShiftX) >> info.chromaShiftX);
+        const uint32_t shiftedH = (uint32_t)((image->height + (uint64_t)info.chromaShiftY) >> info.chromaShiftY);
```

Comment:
`info.chromaShiftX` is an `int`. I had to look [integer promotion rules](https://en.cppreference.com/w/c/language/conversion) up:

> 1) the ranks of all signed integer types are different and increase with their precision
> 2) the ranks of all signed integer types equal the ranks of the corresponding unsigned integer types

So `uint64_t + int` yields `uint64_t`, assuming `sizeof(int) < sizeof(uint64_t)`. It seems to be fine to move the cast but I suggest we leave it as is, for safety and to avoid having to remember the promotion rules.

If I understood correctly, the standard says `uint32_t + int32_t` is implementation defined.

Reply:
Thank you for looking into this. A lot of code (inside and outslde libavif) assumes `int` is the same as `int32_t`. Although strictly speaking this assumption is not portable, it is valid for all the general-purpose CPUs and compilers we need to support today. 

Analyzing this further, we see that `int` is unlikely to be 64 bits, otherwise int16_t and int32_t cannot both be defined as some standard integer types (`short int` is the only one available). It follows that in practice `int` is either 16 bits or 32 bits, and the cast I suggested will result in either `uint64_t + int16_t` or `uint64_t + int32_t`.